server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ChangelogLoader.java

```diff
@@ -0,0 +1,182 @@
+import static java.util.Objects.requireNonNull;
+
+@ServerSide
+public class ChangelogLoader {
+
+  private final DbClient dbClient;
+
+  public ChangelogLoader(DbClient dbClient) {
+    this.dbClient = dbClient;
+  }
+
+  /**
+   * @return non-null list of changes, by descending order of date
+   */
+  public Changelog load(QProfileChangeQuery query) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      List<QProfileChangeDto> dtos = dbClient.qProfileChangeDao().selectByQuery(dbSession, query);
+      List<Change> changes = dtos.stream()
+        .map(Change::from)
+        .collect(Collectors.toList(dtos.size()));
+      completeUserAndRuleNames(dbSession, changes);
+
+      int total = dbClient.qProfileChangeDao().countForProfileKey(dbSession, query.getProfileKey());
+      return new Changelog(total, changes);
+    }
+  }
+
+  private void completeUserAndRuleNames(DbSession dbSession, List<Change> changes) {
+    Set<String> logins = changes.stream().filter(c -> c.userLogin != null).map(c -> c.userLogin).collect(Collectors.toSet());
+    Map<String, String> userNamesByLogins = dbClient.userDao()
+      .selectByLogins(dbSession, logins)
+      .stream()
+      .collect(java.util.stream.Collectors.toMap(UserDto::getLogin, UserDto::getName));
+
+    Set<RuleKey> ruleKeys = changes.stream().filter(c -> c.ruleKey != null).map(c -> c.ruleKey).collect(Collectors.toSet());
+    Map<RuleKey, String> ruleNamesByKeys = dbClient.ruleDao()
+      .selectByKeys(dbSession, Lists.newArrayList(ruleKeys))
+      .stream()
+      .collect(java.util.stream.Collectors.toMap(RuleDto::getKey, RuleDto::getName));
+
+    changes.forEach(c -> {
+      c.userName = userNamesByLogins.get(c.userLogin);
+      c.ruleName = ruleNamesByKeys.get(c.ruleKey);
+    });
+  }
+
+  static class Change {
+    private String key;
+    private String severity;
+    private String userLogin;
+    private String userName;
+    private String type;
+    private String inheritance;
+    private RuleKey ruleKey;
+    private String ruleName;
+    private long at;
+    private final Map<String, String> params = new HashMap<>();
+
+    private Change() {
+    }
+
+    public String getKey() {
+      return key;
+    }
+
+    @CheckForNull
+    public String getSeverity() {
+      return severity;
+    }
+
+    @CheckForNull
+    public String getUserLogin() {
+      return userLogin;
+    }
+
+    @CheckForNull
+    public String getUserName() {
+      return userName;
+    }
+
+    public String getType() {
+      return type;
+    }
+
+    @CheckForNull
+    public String getInheritance() {
+      return inheritance;
+    }
+
+    public RuleKey getRuleKey() {
+      return ruleKey;
+    }
+
+    @CheckForNull
+    public String getRuleName() {
+      return ruleName;
+    }
+
+    public long getCreatedAt() {
+      return at;
+    }
+
+    public Map<String, String> getParams() {
+      return params;
+    }
+
+    private static Change from(QProfileChangeDto dto) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused private "from" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUnusedPrivateMethod)